Doc/c-api/module.rst

```diff
@@ -411,6 +411,31 @@ The available slot types are:
 
    .. versionadded:: 3.12
 
+.. c:macro: Py_mod_gil
+
+   Specifies one of the following values:
+
+   .. c:macro:: Py_MOD_GIL_USED
+
+      The module assumes the presence of the global interpreter lock (GIL), and
+      may access global state without synchronization.
+
+   .. c:macro:: Py_MOD_GIL_NOT_USED
+
+      The module is safe to run without an active GIL.
```

Comment:
Note that there is a lot of overlap with `Py_mod_multiple_interpreters` here.

Per PEP 489, it [defaults to `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`](https://github.com/python/cpython/pull/104148/files#diff-8c2a0fd137780a0ee11c19aae8e46e8490f7667698cd78a5bde466bc995a08c5R315-R317).  That implies a promise that the module relies on module state and doesn't have any static types.

PEP 684 (per-interpreter GIL) added `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`, which is basically one step further: a promise that the module really doesn't have any thread-unsafe state/objects shared between interpreters.  If the module relies on any third-party linked libraries have thread-unsafe state (e.g. openssl) then the module promises to somehow manage that third-party state in a thread-safe way.

Thus, I see `Py_MOD_GIL_USED` as equivalent to `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` and `Py_MOD_GIL_NOT_USED` to `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`.  There is still probably value in having the distinct module slots, but it may help to clarify the relationship in the docs.

Reply:
Hmm, one (or both) of us is misunderstanding this.  Let's see if I can break it down.  The key thing is what users are indicating by using the different slot values.

<details>
<summary>(long analysis)</summary>

----

First of all, in order to use module slots a module must implement PEP 489 multi-phase init.  Let's look at how that relates to various critical features in extension modules.  We focus on features that are impacted by thread-safety and interpreter isolation.  Objects are central to this (along with function execution, of course), but other data is affected too.

|  | single-phase init<BR/>(legacy) | multi-phase init |  | data races | interpreter<BR/>isolation |
| --- | --- | --- | --- | --- | --- |
| module C globals     | allowed  | not allowed   |  | X | X |
| static types               | allowed  | not allowed   |  | X | X |
| other global objects | allowed  | not allowed   |  | X | X |
| external C globals    | allowed   | not allowed? |  | X | X |
| module state            | optional | required        |  | X |  |
| heap types               | optional | required         |  | X |  |
| module functions     | always    | always           |  | X |  |

(Reentrancy is another element to the story, but let's ignore that for now.)

Here's how we deal with data races in the different situations:

|  | single interpreter | legacy<BR/>subinterpreter | isolated subinterpreter<BR>(PEP 684) | single nogil<BR>(PEP 703) | legacy nogil<BR>(PEP 703) | isolated nogil<BR>(PEP 703) |
| --- | --- | --- | --- | --- | --- | --- |
| module C globals     | 1 GIL | 1 GIL | ---       | --- | --- | --- |
| static types               | 1 GIL | 1 GIL | ---       | nogil | nogil | --- |
| other global objects | 1 GIL | 1 GIL | ---       | nogil* | nogil* | --- |
| external C globals    | 1 GIL | 1 GIL | ---       | ---     | ---     | --- |
| module state            | 1 GIL | 1 GIL | N GILs | nogil | nogil | nogil |
| heap types               | 1 GIL | 1 GIL | N GILs | nogil | nogil | nogil |
| module functions     | 1 GIL | 1 GIL | N GILs | nogil | nogil | nogil |

\* nogil will protect objects but cannot protect initializing/changing the global variable that holds an object.

Now let's tie it back to the proposed multi-phase init slots, given the following slot values.

`Py_mod_gil`:

A: `Py_MOD_GIL_USED`  (**default**)
B: `Py_MOD_GIL_NOT_USED`

`Py_mod_multiple_interpreters`:

X: `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED`
Y: `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`  (**default**)
Z: `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`

(All the above are used via PEP 489 multi-phase init, which is specified indicate that the module supports multiple interpreters.)

Here's how each module slot value relates to the GIL:

| slot | value | single<BR/>interpreter | legacy | isolated | single nogil | legacy nogil | isolated nogil |
| --- | --- | --- | --- | --- | --- | --- | --- |
| `Py_mod_gil` |  |  |  |  |  |  |
|  | A | GIL | GIL | GIL | GIL | GIL | GIL |
|  | B | nogil | nogil | nogil | nogil | nogil | nogil |
| `Py_mod_multiple_interpreters` |  |  |  |  |  |  |  |
|  | X | shared | shared | --- | nogil | nogil | --- |
|  | Y | shared | shared | --- | nogil | nogil | --- |
|  | Z | shared | shared | own | nogil | nogil | nogil |

Here's how each relates to the module features from above, where "X" means "allowed":

|  | A | B |  | X | Y | Z |
| --- | --- | --- | --- | --- | --- | --- |
| module C globals     |  |  |  | allowed |  |  |
| static types               |  |  |  | allowed |  |  |
| other global objects |  |  |  | allowed |  |  |
| external C globals     | allowed? |  |  | allowed | ??? |  |
| module state            | required | required |  | required | required | required |
| heap types                | required | required |  | required | required | required |
| module functions     | required | required |  | required | required | required |

----

</details>

Based on the above, here are the permutations and what they mean:

| `Py_mod_multiple_interpreters` | `Py_mod_gil` | module state<BR/>isolated | external state<BR/>isolated | module state<BR/>thread-safe | external state<BR/>thread-safe |
| --- | --- | --- | --- | --- | --- |
| \* | \*\* | X |  |  |  |
| \* | `Py_MOD_GIL_USED` | X |  |  |  |
| \* | `Py_MOD_GIL_USED` | X |  |  |  |
|  |  |  |  |  |  |
| `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` | \*\* |  |  |  |  |
| `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` | \*\* | X |  |  |  |
| `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED` | \*\* | X | X |  | X |
|  |  |  |  |  |  |
| `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` | `Py_MOD_GIL_USED` |  |  |  |  |
| `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` | `Py_MOD_GIL_USED` | X |  |  |  |
| `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED` | `Py_MOD_GIL_USED` | X | X |  | X |
|  |  |  |  |  |  |
| `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` | `Py_MOD_GIL_NOT_USED` |  |  | X | X |
| `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` | `Py_MOD_GIL_NOT_USED` | X |  | X | X |
| `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED` | `Py_MOD_GIL_NOT_USED` | X | X | X | X |

\* defaults to `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`
\*\* defaults to `Py_MOD_GIL_USED`

That's a little bit different from what I had in my head.  `Py_MOD_GIL_NOT_USED` implies at least `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`.  However, it doesn't actually imply `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`, because the module may need to manage external (library) state relative to each interpreter.  (That's an uncommon case, but not one we can ignore.)

`Py_MOD_GIL_USED` doesn't imply anything relative to `Py_mod_multiple_interpreters`.

> the one that feels odd enough that we might want to ban it is Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED with Py_MOD_GIL_NOT_USED

There's no harm in it.  It would just keep the module from being imported in an interpreter that has maximum isolation.  (The GIL is only one element of that level of isolation.)

> I'm guessing we don't want to support the execution mode of "free-threading within an interpreter, but only one interpreter can be active at a time", so what should we do when loading a module with that configuration?

I'm not clear on why we would need to do anything.  Could you elaborate?

Reply:
Thanks for the thorough response!

I'm having trouble making sense of your last table (the one after the collapsed section). Aren't there only 6 possibilities to consider, 3 (`Py_mod_interpreters`) by 2 (`Py_mod_gil`)? Is your table treating "defaults to `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`" differently than a module explicitly providing `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` (and similarly for `Py_MOD_GIL_USED`)?

This is the table I've been using to think about the different options (common prefixes omitted for brevity):
| | `GIL_USED` | `GIL_NOT_USED` |
| --- | --- | --- |
| `MULTIPLE_INTERPRETERS_NOT_SUPPORTED` | uses global state (either in the module itself or a supporting library), not thread-safe | uses global state, but is thread-safe |
| `MULTIPLE_INTERPRETERS_SUPPORTED` | default for multi-phase init (state partitioned by interpreter, not thread-safe) | (discussed below) |
| `PER_INTERPRETER_GIL_SUPPORTED` | state can be accessed by multiple interpreters in parallel, but not more than one thread per interpreter at a time | state can be accessed by multiple interpreters with multiple threads each in parallel |

With that context:

> `Py_MOD_GIL_NOT_USED` implies at least `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`

I don't think this is true. It's not too hard to imagine a C library that is thread-safe but has a global namespace that can't reasonably be used in a way that doesn't share data between subinterpreters (top right square in my table).

And the "(discussed below)" square:

> > the one that feels odd enough that we might want to ban it is Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED with Py_MOD_GIL_NOT_USED
> 
> There's no harm in it. It would just keep the module from being imported in an interpreter that has maximum isolation. (The GIL is only one element of that level of isolation.)

My concern is that if we load such a module in multiple interpreters, while their shared GIL is disabled, we could end up running that module's code in different interpreters in parallel. That seems like the crucial thing that `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` disallows compared to `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`. I chatted with @colesbury about this specific issue earlier today, and he's not concerned about it. If you feel the same way then I'm happy to leave this point alone and move on.

I had been thinking of `Py_MOD_GIL_NOT_USED` as meaning "this module supports free-threading within an interpreter. free-threading across interpreters is determined by the `Py_mod_interpreters` slot." But it sounds like you and Sam have been viewing it as "this module supports free-threading, full stop." Does that framing sound reasonable?

> I'm not clear on why we would need to do anything. Could you elaborate?

This was me working through a more complicated scenario that isn't relevant if the `Py_MOD_GIL_NOT_USED` + `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` combination is freely allowed.